### PR TITLE
YouTrack API

### PR DESCRIPTION
**Pull request by kolbacer**
гет запросы в youtrack

### CODE REVIEW

**Comment by kolbacer** [comment]
.

File: youtrack_api_get/test.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+Gleb_issue_name = "DEMO-20"
+
+status = get_state(Gleb_youtrack_domain, Gleb_issue_name, Gleb_token)
+print(status)
+priority = get_priority(Gleb_youtrack_domain, Gleb_issue_name, Gleb_token)
+print(priority)
+timeEnd = get_time_end(Gleb_youtrack_domain, Gleb_issue_name, Gleb_token)
+print(timeEnd)
+timeStart = get_time_start(Gleb_youtrack_domain, Gleb_issue_name, Gleb_token)
+print(timeStart)
```
Comment (Bezzzub) [inline youtrack_api_get/test.py]: It's not a tests
Rename it to `example.py` and, if you want, write correct test via `pytest`

File: youtrack_api_get/issue_info.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import httpx
+import datetime
+
+state_id = "111-2"
+priority_id = "111-0"
+time_end_id = "117-0"
+
+
+def get_custom_field(youtrack_domain, issue_name, field_id, token):
```
Comment (Bezzzub) [inline youtrack_api_get/issue_info.py]: Return `FieldData` entity, not just string

File: youtrack_api_get/issue_info.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    url = "https://" \
+          + youtrack_domain \
+          + ".myjetbrains.com/youtrack/api/issues/" \
+          + issue_name \
+          + "/customFields/" \
+          + field_id
```
Comment (Bezzzub) [inline youtrack_api_get/issue_info.py]: Common url part move to class attribute.

File: youtrack_api_get/issue_info.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import httpx
+import datetime
+
+state_id = "111-2"
+priority_id = "111-0"
+time_end_id = "117-0"
```
Comment (Bezzzub) [inline youtrack_api_get/issue_info.py]: Move to class attribute

File: youtrack_api_get/issue_info.py (hunk 3 above)
Comment (Bezzzub) [inline youtrack_api_get/issue_info.py]: Use `f-strings`   
f"https://{youtrack_domain}.myjetbrains.com/youtrack/api/issues/{issue_name}\{field_id}"

File: youtrack_api_get/issue_info.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+def get_time_end(youtrack_domain, issue_name, token):
+    return get_custom_field(youtrack_domain, issue_name, time_end_id, token)
+
+
+def get_time_start(youtrack_domain, issue_name, token):
+    url = "https://" \
+               + youtrack_domain \
+               + ".myjetbrains.com/youtrack/api/issues/"\
+               + issue_name
```
Comment (Bezzzub) [inline youtrack_api_get/issue_info.py]: Use `f-strings`

File: youtrack_api_get/issue_info.py (hunk 2 above)
Comment (kolbacer) [reply youtrack_api_get/issue_info.py]: (for me)
class FieldInfo(BaseModel):
    name: str
    value: str

File: youtrack_api_get/issue_info.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    url = "https://" \
+          + youtrack_domain \
+          + ".myjetbrains.com/youtrack/api/issues/" \
+          + issue_name \
+          + "/customFields/" \
+          + field_id
+    headers = {'Authorization': 'Bearer ' + token}
+    params = {'fields': 'projectCustomField(field(name)),value(name)'}
+    response = httpx.get(url, headers=headers, params=params)
+    data = response.json()
```
Comment (Bezzzub) [inline youtrack_api_get/issue_info.py]: Check the response. It is always valid?
```python
try:
    response.raise_for_status()
except HTTPError:
    do something
```

File: youtrack_api_get/issue_info.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+               + ".myjetbrains.com/youtrack/api/issues/"\
+               + issue_name
+    headers = {'Authorization': 'Bearer ' + token}
+    params = {'fields': 'created'}
+    response = httpx.get(url, headers=headers, params=params)
+    data = response.json()
+    name = "Created: "
+    time = data["created"]
+    value = datetime.datetime.fromtimestamp(time//1000.0)
+    return name + ": " + str(value)
```
Comment (Bezzzub) [inline youtrack_api_get/issue_info.py]: Return entity, not just string

File: youtrack_api_get/test.py (hunk 1 above)
Comment (kolbacer) [reply youtrack_api_get/test.py]: done

File: youtrack_api_get/issue_info.py (hunk 2 above)
Comment (kolbacer) [reply youtrack_api_get/issue_info.py]: done

File: youtrack_api_get/issue_info.py (hunk 3 above)
Comment (kolbacer) [reply youtrack_api_get/issue_info.py]: can't move to class attribute, but can into object attribute
(done)

File: youtrack_api_get/issue_info.py (hunk 4 above)
Comment (kolbacer) [reply youtrack_api_get/issue_info.py]: done

File: youtrack_api_get/issue_info.py (hunk 3 above)
Comment (kolbacer) [reply youtrack_api_get/issue_info.py]: done

File: youtrack_api_get/issue_info.py (hunk 5 above)
Comment (kolbacer) [reply youtrack_api_get/issue_info.py]: done

File: youtrack_api_get/issue_info.py (hunk 6 above)
Comment (kolbacer) [reply youtrack_api_get/issue_info.py]: done

File: youtrack_api_get/issue_info.py (hunk 7 above)
Comment (kolbacer) [reply youtrack_api_get/issue_info.py]: done

**Comment by kolbacer** [comment]
**all done, let's merge**

File: youtrack_api/IssueManager.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    def _get_custom_field(self, field_id):
+        url = f"{self.base_url}/customFields/{field_id}"
+        headers = {'Authorization': f"Bearer {self.token}"}
+        params = {'fields': 'projectCustomField(field(name)),value(name)'}
+        response = httpx.get(url, headers=headers, params=params)
+        try:
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
+            return FieldInfo(None, None, False)
```
Comment (Bezzzub) [inline youtrack_api/IssueManager.py]: Возвращай None

File: youtrack_api/IssueManager.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    def get_time_end(self):
+        url = f"{self.base_url}/customFields/{IssueManager.time_end_id}"
+        headers = {'Authorization': f"Bearer {self.token}"}
+        params = {'fields': 'projectCustomField(field(name)),value(presentation)'}
+        response = httpx.get(url, headers=headers, params=params)
+        try:
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
```
Comment (Bezzzub) [inline youtrack_api/IssueManager.py:47]: Return None

File: youtrack_api/IssueManager.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    def get_time_start(self):
+        url = self.base_url
+        headers = {'Authorization': f"Bearer {self.token}"}
+        params = {'fields': 'created'}
+        response = httpx.get(url, headers=headers, params=params)
+        try:
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
+            return FieldInfo(None, None, False)
```
Comment (Bezzzub) [inline youtrack_api/IssueManager.py]: None

File: youtrack_api/IssueManager.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            "value": {
+                "name": priority  # Возможные значения приоритета: Show-stopper, Critical, Major, Normal, Minor
+            }
+        }
+        data = json.dumps(data, indent=4)
+        response = httpx.post(url, headers=headers, params=params, data=data)
+        try:
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
```
Comment (Kondrahin) [inline youtrack_api/IssueManager.py:90]: None

File: youtrack_api/IssueManager.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+                "presentation": time  # Оценочное время в формате 1н 1д 1ч 1м
+            }
+        }
+        data = json.dumps(data, indent=4)
+        response = httpx.post(url, headers=headers, params=params, data=data)
+        try:
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
+            return FieldInfo(None, None, False)
```
Comment (Bezzzub) [inline youtrack_api/IssueManager.py]: Return None

File: youtrack_api/IssueManager.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+                "name": priority  # Возможные значения приоритета: Show-stopper, Critical, Major, Normal, Minor
+            }
+        }
+        data = json.dumps(data, indent=4)
+        response = httpx.post(url, headers=headers, params=params, data=data)
+        try:
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
+            return FieldInfo(None, None, False)
```
Comment (Bezzzub) [inline youtrack_api/IssueManager.py]: Return None

File: youtrack_api/schemas/FieldInfo.py [hunk 14]
```diff
@@ -0,0 +1,11 @@
+class FieldInfo:
+    def __init__(self, name, value, is_valid=True):
```
Comment (Bezzzub) [inline youtrack_api/schemas/FieldInfo.py]: Is_valid не нужно.

File: youtrack_api/schemas/FieldInfo.py [hunk 15]
```diff
@@ -0,0 +1,11 @@
...
+        self.name = name  # always string
+        self.value = value  # not always string
+        self.is_valid = is_valid
+
+    def show(self):
```
Comment (Bezzzub) [inline youtrack_api/schemas/FieldInfo.py]: Это тоже не нужно

File: youtrack_api/schemas/FieldInfo.py [hunk 16]
```diff
@@ -0,0 +1,11 @@
+class FieldInfo:
```
Comment (Bezzzub) [inline youtrack_api/schemas/FieldInfo.py]: Лучше отнаследоваться от `BaseModel` из [Pydantic](https://pydantic-docs.helpmanual.io/)
```python
from pydantic import BaseModel

class FieldInfo(BaseModel)
…
```

File: youtrack_api/IssueManager.py (hunk 8 above)
Comment (kolbacer) [reply youtrack_api/IssueManager.py]: done

File: youtrack_api/IssueManager.py (hunk 9 above)
Comment (kolbacer) [reply youtrack_api/IssueManager.py:47]: done

File: youtrack_api/IssueManager.py (hunk 10 above)
Comment (kolbacer) [reply youtrack_api/IssueManager.py]: done

File: youtrack_api/IssueManager.py (hunk 12 above)
Comment (kolbacer) [reply youtrack_api/IssueManager.py]: DONE

File: youtrack_api/IssueManager.py (hunk 13 above)
Comment (kolbacer) [reply youtrack_api/IssueManager.py]: DOOONNNEEEE

File: youtrack_api/schemas/FieldInfo.py (hunk 14 above)
Comment (kolbacer) [reply youtrack_api/schemas/FieldInfo.py]: ok

File: youtrack_api/schemas/FieldInfo.py (hunk 15 above)
Comment (kolbacer) [reply youtrack_api/schemas/FieldInfo.py]: ок

File: youtrack_api/schemas/FieldInfo.py (hunk 16 above)
Comment (kolbacer) [reply youtrack_api/schemas/FieldInfo.py]: сделал

**Review by Bezzzub — APPROVED** [review]

**Pull request merged**